### ISSUE TITLE
Variable expansion not working?

### ISSUE THREAD

**Issue by mathsen** [coreruleset/nextcloud-rule-exclusions-plugin#2]
### Describe the bug
I have a nextcloud instance and just switched from the "old" exclusions to the new plugin system.
During the tests I reckognized a problem:
[Rule 9508130](https://github.com/coreruleset/nextcloud-rule-exclusions-plugin/blob/65015038c617f58d6fe63a6c33e983f6dda7b21e/plugins/nextcloud-rule-exclusions-before.conf#L202) should expand tx.allowed_methods:
````
setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT PATCH CHECKOUT COPY DELETE LOCK MERGE MKACTIVITY MKCOL MOVE PROPFIND PROPPATCH SEARCH UNLOCK REPORT TRACE jsonp'"`
````
tx.allowed_methods is the default value (GET HEAD POST OPTIONS), so I don't set it - but somehow it is not expanded here.
I receive the following block in my log:

````
---xbaiLTEY---A--
[29/Apr/2022:17:06:53 +0200] XXXX
---xbaiLTEY---B--
GET /index.php/avatar/xxx/384 HTTP/1.1
[...]
Host: XXX
User-Agent: Mozilla/5.0 (Android) Nextcloud-android/3.19.1
[...]


---xbaiLTEY---F--
HTTP/1.1 403
Date: Fri, 29 Apr 2022 15:06:53 GMT
Content-Length: 146
Content-Type: text/html
Connection: keep-alive

---xbaiLTEY---H--
ModSecurity: Warning. Matched "Operator `Within' with parameter ` PUT PATCH CHECKOUT COPY DELETE LOCK MERGE MKACTIVITY MKCOL MOVE PROPFIND PROPPATCH SEARCH UNLOCK REPORT TRACE jsonp' against variable `REQUEST_METHOD' (Value: `GET' ) [file "/etc/nginx/modsec/coreruleset-dev/rules/REQUEST-911-METHOD-ENFORCEMENT.conf"] [line "28"] [id "911100"] [rev ""] [msg "Method is not allowed by policy"] [data "GET"] [severity "2"] [ver "OWASP_CRS/4.0.0-rc1"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/210/272/220/274"] [tag "PCI/12.1"] [hostname "XXX"] [uri "XXXX"] [unique_id "1651244813"] [ref "v0,3"]
ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:BLOCKING_INBOUND_ANOMALY_SCORE' (Value: `5' ) [file "/etc/nginx/modsec/coreruleset-dev/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "176"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"] [data ""] [severity "0"] [ver "OWASP_CRS/4.0.0-rc1"] [maturity "0"] [accuracy "0"] [tag "anomaly-evaluation"] [hostname "XXXX"] [uri "XXXX"] [unique_id "1651244813"] [ref ""]
````

So in the error all the methods, which are extended to tx.allowed_methods are listed (PUT PATCH CHECKOUT COPY DELETE LOCK MERGE MKACTIVITY MKCOL MOVE PROPFIND PROPPATCH SEARCH UNLOCK REPORT TRACE jsonp), but not the default ones.

### Steps to reproduce
Installed CRS with nextcloud plugin, access a shared folder via Android App

### Expected behaviour
Rule 9508130 should not overwrite but add methods.

### Actual behaviour
Rule 9508130 should add the methods correctly.

### Your Environment

* CRS version: 4.0.0-rc1
* Paranoia level setting:1
* ModSecurity version: 3.0.6
* Web Server and version: nginx 1.21.6
* Operating System and version: Arch Linux

Maybe someone can help me with this?

//Edit:
I just realized, that I have this issue also with other rules, e.g. 9508110 expands tx.allowed_request_content_type - and there I have the same issue.

**Comment by theseion** [comment]
Thanks for reporting @mathsen. I've moved the issue to the nextcloud plugin repository.

**Comment by azurit** [comment]
Can you change `phase` to `2` for rule `9508130` and try again?

**Comment by mathsen** [comment]
Hello,

changing the phase to 2 doesn't change the result (I'm blocked) but now because of a PROPFIND request:

````
---s0e62XLU---A--
[30/Apr/2022:08:22:00 +0200] 1651299720 XXXX 443
---s0e62XLU---B--
PROPFIND /remote.php/dav/files/XXX/ HTTP/1.1
Depth: 0
Host: XXX
User-Agent: Mozilla/5.0 (Android) Nextcloud-android/3.19.1
Content-Length: 466
Content-Type: text/xml; charset=UTF-8

[...]

---s0e62XLU---H--
ModSecurity: Warning. Matched "Operator `Within' with parameter `GET HEAD POST OPTIONS' against variable `REQUEST_METHOD' (Value: `PROPFIND' ) [file "/etc/nginx/modsec/coreruleset-dev/rules/REQUEST-911-METHOD-ENFORCEMENT.conf"] [line "28"] [id "911100"] [rev ""] [msg "Method is not allowed by policy"] [data "PROPFIND"] [severity "2"] [ver "OWASP_CRS/4.0.0-rc1"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/210/272/220/274"] [tag "PCI/12.1"] [hostname "XXX"] [uri "/remote.php/dav/files/XXX/"] [unique_id "1651299720"] [ref "v0,8"]
ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:BLOCKING_INBOUND_ANOMALY_SCORE' (Value: `5' ) [file "/etc/nginx/modsec/coreruleset-dev/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "176"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 5)"] [data ""] [severity "0"] [ver "OWASP_CRS/4.0.0-rc1"] [maturity "0"] [accuracy "0"] [tag "anomaly-evaluation"] [hostname "XXX"] [uri "/remote.php/dav/files/XXX/"] [unique_id "1651299720"] [ref ""]

[...]
````

So I guess the adjustment of 9508130 is too late now - this is a request, where this rule should add PROPFIND as an allowed request method.

**Comment by theseion** [comment]
Hm. Maybe we can put those rules into the `*-after.conf` file @azurit?

**Comment by theseion** [comment]
I think that should work...

@mathsen would you mind giving it a try? Create a file called `nextcloud-rule-exclusions-after.conf` and move the phase 1 rules that set transaction variables to that file (9508600, 9508110, 9508115, 9508130, 9508140, 9508320, 9508330, 9508601, 9508602, 9508603, 9508701, 9508850).
Inclusion should happen automatically.

**Comment by azurit** [comment]
> Hm. Maybe we can put those rules into the `*-after.conf` file @azurit?

That's an alternative solution. Maybe better in this case.

**Comment by azurit** [comment]
> I think that should work...
> 
> @mathsen would you mind giving it a try? Create a file called `nextcloud-rule-exclusions-after.conf` in the plugins directory and move the phase 1 rules that set transaction variables to that file (9508600, 9508110, 9508115, 9508130, 9508140, 9508320, 9508330, 9508601, 9508602, 9508603, 9508701, 9508850). Inclusion should happen automatically.

@mathsen Or just rename file `nextcloud-rule-exclusions-before.conf` to `nextcloud-rule-exclusions-after.conf`.

**Comment by mathsen** [comment]
@azurit 
Changing the filename to `nextcloud-rule-exclusions-after.conf` didn't change anything (I also reverted the change to phase 2 from before, so I used the current file of the 4.0.0-rc1 tag) and renamed it.
I get the [same block as before](https://github.com/coreruleset/nextcloud-rule-exclusions-plugin/issues/2#issuecomment-1113929636).
Just to be sure, I include the files in the following order:
````
Include /etc/nginx/modsec/modsecurity.conf
Include /etc/nginx/modsec/coreruleset-dev/crs-setup.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-config.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-before.conf
Include /etc/nginx/modsec/coreruleset-dev/rules/*.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-after.conf
````

**Comment by mathsen** [comment]
Hello again,
I have made some debugging and the behavior of modsecurity / CRS with this plugin is really strange.
I made four experiments:

1) rule 9508130 in phase 1, putting all rules in nextcloud-rule-exclusions-before.conf
2) rule 9508130 in phase 2, putting all rules in nextcloud-rule-exclusions-before.conf
3) rule 9508130 in phase 1, putting all rules in nextcloud-rule-exclusions-after.conf
4) rule 9508130 in phase 2, putting all rules in nextcloud-rule-exclusions-after.conf

Here are the relevant result:
[logs.zip](https://github.com/coreruleset/nextcloud-rule-exclusions-plugin/files/8597602/logs.zip)

Basically tx.allowed_methods is not set or it is set exclusively by the nextcloud plugin without using the old value of %{tx.allowed_methods}.
It's really strange to me - nextcloud is basically unusable for me with CRS 4.0.0-rc1 and I don't know what's going wrong.

**Comment by mathsen** [comment]
I have another addition to make:
I basically use the `crs-setup.conf` [default file](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/crs-setup.conf.example).
Uncommenting [rule 900200](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/crs-setup.conf.example#L427) fixes the above issue for me - then `tx.allowe_methods` is set correctly, when rule 9508130 of the nextcloud plugin is executed and the above error with this plugin is gone in it's "default state" (rules in phase 1 ande *-before.conf).

I get similar errors with tx.allowed_request_content_type - enabling [rule 900220](https://github.com/coreruleset/coreruleset/blob/911b8c1d5d74713f6495f21788a912edd3ca65e2/crs-setup.conf.example#L446) also fixes this issue for me.

So whyever these variables are not set when executing the plugin when they are outcommented (as it is in the example file). I assumed, that when I don't set them manually, that they have their default value?

So with the above two uncomments the nextcloud app is basically usable for me - I didn't do any further debugging for now - can anybody give an explanation for this?

**Comment by mathsen** [comment]
I have further debugged the problem and I guess I found the underlying problem. I realized it, when I reamed the nextcloud rules to *-after.conf. This resulted in the same error as before: 911100 blocks my (valid) requests.
Further I saw that rule 901160 of `REQUEST-901-INITIALIZATION.conf` basically does the same as my manually uncommented rule in crs-setup.conf.

So my conclusion for now is as followings:

This include order:
````
Include /etc/nginx/modsec/modsecurity.conf
Include /etc/nginx/modsec/coreruleset-dev/crs-setup.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-config.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-before.conf
Include /etc/nginx/modsec/coreruleset-dev/rules/*.conf
Include /etc/nginx/modsec/coreruleset-dev/plugins/*-after.conf
````

leads in a basic setup to the problem, that `plugins/*-before.conf` gets evaluated before the CRS default values are set (happens in `rules/REQUEST-901-INITIALIZATION.conf`).
Therefore I got the error in my first post.

Putting all the rules to `plugins/*-after.conf` leads to the problem, that all `rules/*.conf` are processed and executed before the plugin - and during this processing the requests, that should be allowed in `plugins/*-after.conf` are already blocked (or the anomaly score is increased).

So as I see it: the problem is, that the blocking (or incrasing of the anomaly score) already happens in `rules/*.conf`, before taking the changes of the `plugins/*-after.conf` in consideration.

I assume, that this was never a problem before, because in [CRS < 4.0](https://github.com/coreruleset/coreruleset/tree/v3.3.2/rules) the nextcloud rules (`rules/REQUEST-903.9003-NEXTCLOUD-EXCLUSION-RULES.conf`) were simply included before `rules/REQUEST-911-METHOD-ENFORCEMENT.conf` (the „showstopper“ in my case).

So in a way I would say in it's current state the system is broken „by design“. The plugin files should be evaluated after `rules/REQUEST-901-INITIALIZATION.conf` but before all the other rules - which is not possible with the current system.

**Comment by azurit** [comment]
@mathsen Yes, you are right. We will think about it.

**Comment by dune73** [comment]
Thank you for debugging this @mathsen. You got to the root of it.

I kind of designed this and I know of the shortcoming, but I did not expect one of the plugins to suffer from it. But that's why we do release-candidates after all. Thank you for giving it a go, so we have time to fix this.

It is a bit hairy and will take some thinking.

**Comment by dune73** [comment]
Here is a proposal for a potential solution:

The plugin does the same as the init in 901 itself. When tx.allowed_methods is not set, then it sets it to the default in a rule like 9508090.

So if you set it in crs-setup.conf, then everything works as planned.
If you do not set it in crs-setup.conf and the plugin runs with an uninitialized allowed_methods, we set it to the default and everything works as planned.

@azurit, @mathsen, @theseion : please think this through and confirm if you think it would work. Experimenting with this also welcome of course.

**Comment by RedXanadu** [comment]
@dune73 This means that if the "default" value for such a variable is changed in the future (in this specific case, `tx.allowed_methods`) then it must be changed in the CRS (`crs-setup.conf` and `REQUEST-901-INITIALIZATION.conf`) *and* changed in every plugin that follows this proposed construct of "set the default value if the variable doesn't already exist".

Maybe cross-referencing in the comments would be a good idea? E.g. in `crs-setup.conf` we could say: "tx.allowed_methods is also defined in the following files and official plugins: …"

**Comment by dune73** [comment]
I guess that will be the case. And if we forget, then it's an issue that somebody will stumble over.

Such a comment in crs-setup.conf is worth a consideration. But it also means more things to think about and more interdependence between plugins and mainline.

**Comment by dune73** [comment]
Also: Changing default http methods is probably very, very rare and when we do it's quite obvious we would think of browsing the plugins for where the value is being referenced.

**Comment by mathsen** [comment]
@dune73 Thanks for the clarification and the idea to fix it.
I'm not a developer of CRS and therefore have a lack of deeper knowledge of the single rules. But from my perspective the defaults should be set after the initialization of CRS and before processing the "real" rules.
That could maybe be achieved, by including `REQUEST-901-INITIALIZATION.conf` at the very end in crs-setup.conf and move it out of the `rules` folder?

One thing, that speaks against this, is if a rule in `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` assumes that no default values are set - then these rules would break, of course.
I personally don't have such a rule, but I don't know if one can exclude such a case generally.

Such an approach would have the advantage, that one doesn't need code duplication (default values are set in one place and don't have to be re-set in a plugin file, if it modifies a TX value).

What do others think about this? Maybe there is a drawback, which I just don't see.

**Comment by mathsen** [comment]
Hello,

I just wanted to ask if there is any progress on this matter (so that the issue will be resolved with the final version 4.0?).
As non developer of CRS, I don't know if I have a vote, but personally I would go with approach #2 of @azurit, since less include statements are necessary than with #1 and just looks "simpler" to me.

Of course with #1 one is more flexible, but in my opinion it makes only sense, if `plugins/*-config.conf` or `plugins/*-before.conf` rely on rules, that have to be executed before default values are set.
I just can't imagine such a scenario (which doesn't mean it's not possible), but even with CRS 3.3.2 no plugin executed a rule before default values were set.

I still have another concern: with 3.3.2 `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` was evaluated before default values were set. So with all approaches of @azurit this rule will then executed after default values are set.
Again: this is no big deal from my point of view, but if any user relies on no set default values, this file has explicitly to be included before default rules are set.

**Comment by dune73** [comment]
We have not decided on this @mathsen. In fact there is a private bug bounty running and it is delaying our CRS v4 planning with a lot of new issues being reported. So we are a bit absorbed but we will return to this in time.

**Comment by chriswells0** [comment]
This thread has a few debugging summaries and several idea/suggestions, but I'm not sure from reading through it what a current workaround would be (not a CRS expert here). Is it possible for this plug-in to be used today with a workaround? If so, which one or what are the steps? I'm using CRS 3.4, if that makes any difference, but I don't mind upgrading to the 4.0 dev branch.

If it can't be used yet, would it be better to go back to 3.3 with REQUEST-903.9003-NEXTCLOUD-EXCLUSION-RULES.conf?

**Comment by theseion** [comment]
Hi @chriswells0,
Yes, it's possible to use, you just have to make sure that you set `tx.allowed_methods` properly in the plugin. We may have to modify the whole plugin system a bit to fix the core issue, which is initialization order.
And yes, it is possible to use the plugin with 3.4, just follow the instructions in the [docs](https://coreruleset.org/docs/concepts/plugins/#how-to-install-a-plugin).

We would be *very* happy, if you would test v4.0/dev for us, of course, but if you're running a production system, then it would probably be better to wait. We're currently in Italy at the retreat and are making a _lot_ of changes to v4.0/dev.

**Comment by chriswells0** [comment]
Thanks a ton for such a quick response. What is the proper way to set `tx.allowed_methods`? I see this commented out in _crs-setup.conf_ and uncommented in _rules/REQUEST-901-INITIALIZATION.conf_:
```
setvar:'tx.allowed_methods=GET HEAD POST OPTIONS'"
```
Is it simply to uncomment that line in _crs-setup.conf_?

Is it possible to apply a plug-in to only 1 virtual host in Apache while using CRS globally? It seems with the way the rules are included between "before" and "after" plugin files that there's no way to have rules included globally and plugin files in 1 virtual host. Is that right or is it OK to re-include the rules in a virtual host?

**Comment by theseion** [comment]
At the top of your `nextcloud-rule-exclusions-before.conf`, add this:

```
# Default HTTP policy: allowed_methods (rule 900200 in crs-setup.conf)
SecRule &TX:allowed_methods "@eq 0" \
    "id:9508090,\
    phase:1,\
    pass,\
    nolog,\
    ver:'OWASP_CRS/4.0.0-rc1',\
    setvar:'tx.allowed_methods=GET HEAD POST OPTIONS'"
```

That's a copy of rule 901160 from `setup.conf` that will initialize the variable to its default if it hasn't been set already. That should be all.

**Comment by UM-Li** [comment]
Hi, rule 911100 still gives false positives on Nextcloud v25.0.3.  This breaks file preview in the WebUI.

**Comment by azurit** [comment]
@UM-Li Modify rule above like this:
```
...
    setvar:'tx.allowed_methods=GET HEAD POST OPTIONS PUT'"
```

**Comment by theseion** [comment]
This issue should be solved after having merged #14. See https://github.com/coreruleset/coreruleset/issues/3154 for reference.
